dace/libraries/blas/utility/memoryOperations.py

```diff
@@ -0,0 +1,299 @@
+    outputs = []
+    names = []
+
+    for dest, size, dtype, i in inputs:
+
+        src = "fpga_" + dest
+
+        name, desc = sdfg.add_array(
+            src,
+            shape=[size],
+            dtype=dtype,
+            storage=dtypes.StorageType.FPGA_Global,
+            transient=True
+        )
+        if bank is not None:
+            if isinstance(bank, list):
+                desc.location = {"bank": bank[i]}
+            else:
+                desc.location = {"bank": bank}
+            print("Set bank of:", name, " to ", desc.location)
+
+        fpga_in = state.add_read(src)
+        cpu_out = state.add_write(dest)
+
+        state.add_memlet_path(
+            fpga_in, cpu_out,
+            memlet=Memlet.simple(cpu_out.data, "0:{}".format(size))
+        )
+
+        outputs.append(fpga_in)
+        names.append(src)
+
+    return (outputs, names)
+
+
+def fpga_copyGlobalToLocalSubset(state, src, dest, size, start):
+
+    x_in = state.add_read(src)
+    x_buf = state.add_write(dest)
+
+    copyMap_entry, copyMap_exit = state.add_map(
+        'copyX_map',
+        dict(k = '{0}:{0}+{1}'.format(start, size)),
+        schedule=dtypes.ScheduleType.FPGA_Device,
+        unroll=True
+    )
+
+    copyX_task = state.add_tasklet(
+        'copyX_task',
+        ['inCon'],
+        ['outCon'],
+        'outCon = inCon'
+    )
+
+    state.add_memlet_path(
+        x_in, copyMap_entry, copyX_task,
+        dst_conn='inCon',
+        memlet=Memlet.simple(x_in.data, "k")
+    )
+
+    state.add_memlet_path(
+        copyX_task, copyMap_exit, x_buf,
+        src_conn='outCon',
+        memlet=Memlet.simple(x_buf.data, "k - {}".format(start))
+    )
+
+
+def fpga_copyGlobalToLocalTile(state, src, dest, rowSize, rowStart, colSize, colStart, globalCol):
+
+    x_in = state.add_read(src)
+    x_buf = state.add_write(dest)
+
+    copyMap_entry, copyMap_exit = state.add_map(
+        'copyGlobalToLocal_map_{}'.format(dest),
+        dict(k_row = '{0}:{0}+{1}'.format(rowStart, rowSize)),
+        schedule=dtypes.ScheduleType.FPGA_Device
+    )
+
+    copyMapCol_entry, copyMapCol_exit = state.add_map(
+        'copyGlobalToLocalCol_map_{}'.format(dest),
+        dict(k_col = '{0}:{0}+{1}'.format(colStart, colSize)),
+        schedule=dtypes.ScheduleType.FPGA_Device,
+        unroll=True
+    )
+
+
+    copyX_task = state.add_tasklet(
+        'copyX_task',
+        ['inCon'],
+        ['outCon'],
+        'outCon = inCon'
+    )
+
+    state.add_memlet_path(
+        x_in, copyMap_entry,copyMapCol_entry, copyX_task,
+        dst_conn='inCon',
+        memlet=Memlet.simple(x_in.data, "k_row * {} + k_col".format(globalCol))
+    )
+
+    state.add_memlet_path(
+        copyX_task, copyMapCol_exit, copyMap_exit, x_buf,
+        src_conn='outCon',
+        memlet=Memlet.simple(x_buf.data, "(k_row - {0}) * {1} + (k_col - {2})".format(rowStart, colSize, colStart))
+    )
+
+
+def fpga_streamToLocal(state, srcData, dest, size):
```

Comment:
I had issues with the direct memlet connection at some point and therefore wrote it like this. It is being used in quite a lot of places now and I would appreciate it, if I can keep it like this.

Reply:
Could you please check if it "just" works if you replace it with a container-to-container memlet?

If it still causes problems, then we can keep it like this for now.

The reason why I would prefer container-to-container is that it's more robust towards changes that we make in DaCe. For example, when we introduce the vector types, it will require zero changes to your SDFG if you have a container-to-container copy.